[게시글 생성/조회/수정/삭제] 러너덕 미션 제출합니다.

Deocksoo opened the pull request:
현구님 안녕하세요, 이번 미션 리뷰이 러너덕입니다 :)
이번에 Spring을 처음 접해봐서 아직 코드에 부족한 부분이 많은 것 같아요.
페어와 함께 미션을 구현하면서 모호했던 부분도 많았는데, 그 중 두 가지를 정리해 보았어요. 
괜찮으시다면 혹시 의견 부탁드려도 될까요?

1. Repository에 의존성이 있는 ArticleController를 어떻게 테스트해야 할 지 조금 모호했습니다. 그래서 다음과 같은 두 가지 방법을 생각해봤는데요, 
- 테스트 메서드를 실행할 때마다 webTestClient로 서버에 요청을 해서 repository의 상태를 바꿔주는 방법
- repository를 test에도 주입해서 직접 repository의 상태를 바꿔주는 방법 (제출한 코드에서 사용한 방식)

   실제로는 이렇게 의존성을 가진 Controller들을 어떻게 테스트하는지 궁금해요. 혹시 위 두 가지 방법  
   이외에 다른 좋은 방법이 있을까요?

2. Controller에서 Url을 mapping 할 때, Path와 같은 클래스를 만들어서 각각의 Url들을 String 상수로 관리해보면 어떨까 하는 생각이 들었습니다. 그런데 또 이렇게 했다가 path variable들이 보이지 않으면 되려 가독성이 떨어질 수도 있을 것 같아 고민이 됩니다. 이 부분에 대해 어떻게 생각하시나요?

구현한 ArticleRepository도 너무 단순한 것 같아서, 이렇게 구현해도 되는지 잘 모르겠네요.
수정해야할 사항들을 말씀해 주시면 적극 고민해보겠습니다!
읽어주셔서 감사합니다 :)

Deocksoo committed:
[DOCS] 해야될 일 목록을 업로드합니다

Deocksoo committed:
[DOCS] 기능 목록을 업데이트합니다

Deocksoo committed:
[ADD] 프로젝트에 필요한 의존성을 추가합니다

Deocksoo committed:
[ADD] Article과 ArticleRepository 도메인 객체를 추가합니다

Deocksoo committed:
[TEST] ArticleRepository 테스트에 주석을 추가합니다

Deocksoo committed:
[FIX] app.css 오류 코드를 수정합니다

Deocksoo committed:
[ADD] 게시글 작성 후 작성된 게시글을 보여주는 기능을 추가합니다

1. Article이 생성될 때 데이터를 저장하도록 수정하였습니다.

Deocksoo committed:
[ADD] IndexController를 추가합니다

index 페이지를 thymeleaf 템플릿 엔진에서 templates 디렉토리를 base로 하여 찾아줄 수
있도록 수정하였습니다.

Deocksoo committed:
[ADD] 메인 페이지 게시글 목록 조회 기능을 추가하였습니다

Deocksoo committed:
[ADD] 저장한 게시물을 조회하는 기능을 구현합니다

Deocksoo committed:
[REFACTOR] ArticleRepository find 메소드 리팩토링

Deocksoo committed:
[REFACTOR] article.html css, js 경로 수정하였습니다

Deocksoo committed:
[ADD] HellowWordController 실습 파일을 업로드합니다

Deocksoo committed:
[TEST] IndexController에서 사용하지 않는 라이브러리를 제거합니다

Deocksoo committed:
[ADD] 게시글 수정 및 해당 페이지로 이동하는 기능을 추가하였습니다

Deocksoo committed:
[REFACTOR] 컨벤션에 따른 수정 및 불필요한 라이브러리를 제거하였습니다

Deocksoo committed:
[ADD] 게시물 update 기능을 추가합니다

Deocksoo committed:
[ADD] 게시글 삭제 기능 구현 하였습니다

Deocksoo committed:
[REFACTOR] Article 클래스의 field 변수를 수정합니다

Deocksoo committed:
[REFACTOR] Custom exception class들을 추가합니다

Deocksoo committed:
[REFACTOR] ArticleController 메서드들의 이름과 순서를 변경합니다

Deocksoo committed:
[REFACTOR] Article.html파일의 script에서 initialValue를 수정합니다

Thymeleaf 템플릿 엔진을 통해 초기값을 지정해주도록 수정하였습니다.

Deocksoo committed:
[DOCS] 기능 목록을 업데이트합니다

Deocksoo committed:
[FIX] Custon Exception의 접근제어자를 변경합니다

Deocksoo committed:
[REFACTOR] Template 파일들을 리펙토링합니다

1. Template 파일들의 중복되는 코드들을 추출하였습니다.
2. 좌측 상단 배너를 클릭하면 index 페이지로 이동하도록 수정하였습니다.

Deocksoo committed:
[TEST] ArticleControllerTest의 articlePost메서드를 리펙토링 합니다

테스트에 사용할 지역변수들의 이름과 테스트 값을 변경하였습니다. input 값을 한글로
작성하는 경우 encoding 문제로 인해 성공하는 케이스가 실패하는 버그가 발생합니다.
이를 방지하기 위해 영어로 input 값들을 변경하였습니다.

Deocksoo committed:
[TEST] 몇몇 Test 클래스들의 어미를 Tests에서 Test로 변경합니다

Deocksoo committed:
[TEST] 불필요한 MyblogApplicationTest를 제거합니다

Deocksoo committed:
[REFACTOR] ArticleRepositoryTest를 리펙토링 합니다

1. 필드 상수명을 변경하였습니다.
2. 메서드 순서를 정리하였습니다.
3. 수정, 삭제할 때 id가 없는 경우에 대한 테스트 케이스들을 추가하였습니다.

Deocksoo committed:
[TEST] HelloWorldControllerTest에 DisplayName을 추가합니다

Deocksoo committed:
[REFACTOR] ArticleControllerTest를 리펙토링합니다

1. 필드 상수명을 변경하였습니다
2. 테스트 메서드명을 정리하였습니다.
3. 불필요한 테스트를 제거하였습니다.

Deocksoo committed:
[REFACTOR] ArticleRepository에서 Id를 할당하는 방식을 변경합니다

기존의 AutoIncrement 방식에서, 외부에서 id를 설정해 저장해주는 방식으로 변경하였습니다.

Deocksoo committed:
[ADD] ArticleController에서 수정, 삭제 후 redirect하도록 수정하였습니다

Deocksoo committed:
[FIX] index.html boarder 선이 스크롤 따라 움직이는 버그를 수정하였습니다

Deocksoo committed:
[ADD] 업데이트 된 css 파일을 추가합니다

Deocksoo committed:
[REFACTOR] ArticleController가 의존성을 주입받는 위치를 변경합니다

1. ArticleRepository를 필드변수로 주입받던 기존 방식에서 생성자 주입 방식으로
변경하였습니다.
2. 필드 변수의 불필요한 접근제어자를 변경하였습니다.

Deocksoo committed:
[FIX] 게시물의 coverUrl이 index페이지에 노출되지 않는 버그를 수정합니다

Deocksoo committed:
[REFACTOR] Template 파일들의 디렉토리를 static에서 templates로 변경합니다

Deocksoo committed:
[TEST] ArticleRepositoryTest의 오타를 수정합니다

Deocksoo committed:
[REFACTOR] ArticleRepository의 데이터 접근 메서드들과 필드 변수명을 수정합니다

kang-hyungu reviewed (changes requested):
1. 지금 컨트롤러에서 한 테스트는 인수 테스트(acceptance test)로 볼 수 있습니다. 인수 테스트는 실제 운영 환경에서 사용할 준비가 되었는지 확인하는 테스트입니다. 즉, 개발자가 개입하지 않고 개발된 api나 화면을 직접 실행해보면서 문제가 없는지 점검할 수 있어야합니다.
repository 빈을 생성해서 값을 주입하게 되면 개발자가 운영 환경에 개입한게 되겠죠. 개발한 결과물이 정상 동작하는지 확인하려면 webTestClient를 사용해서 테스트해보세요.
그리고 인수 테스트가 아닌 상황에서 의존성이 있는 객체들을 테스트할 때는 mock을 사용하면 됩니다. 요건 차차 배워가실거에요.

2. uri를 문자열 상수로 관리하면 어떤 장점이 있나요? uri는 가독성 높은게 더 중요한 것 같네요.
복잡하게 생각하면 코드도 복잡해져요. 때로는 단순하게 작성하는게 더 좋습니다.

Deocksoo committed:
[REFACTOR] CouldNotFindArticleIdException이 RuntimeException을 상속하도록 수정합니다

Deocksoo committed:
[REFACTOR] IndexController에서 생성자 주입을 받도록 수정합니다

Deocksoo committed:
[REFACTOR] validator 패키지의 이름을 exception으로 변경합니다

Deocksoo committed:
[REFACTOR] Controller들이 의존하는 repository의 불변성을 보장하도록 수정합니다

Deocksoo committed:
[REFACTOR] ArticleRepository의 save 메서드가 ArticleDTO를 인자로 받도록 수정합니다

1. Article Id 관리를 Controller가 아닌 Repostory에서 해주도록 하기 위해 Repository 내부에서
Autoincrement 방식으로 id를 넘버링하도록 수정하였습니다.
2. ArticleDTO를 통해 데이터를 입력받고, Repository에 전달해주도록 수정하였습니다. save 메서드를
이렇게 수정하면 Controller에서 담당하던 Article 객체의 생성을 Repository로 넘길 수 있게 됩니다.

Deocksoo committed:
[REFACTOR] ArticleController의 createNewArticle, editArticle 메서드 인자를 변경합니다.

1. 더 많은 데이터를 넘겨받는 경우를 고려하여 DTO를 통해 데이터를 입력받도록 수정하였습니다.
2. editArticle 메서드에 대한 테스트 케이스가 누락된 것을 발견하여 이를 추가하였습니다.

Deocksoo committed:
[REFACTOR] Article의 setter를 제거합니다

setter를 제거하고, 내용을 update하는 경우 새로운 article을 만들어서 update 하도록
AritcleRepostiroy를 수정하였습니다.

Deocksoo committed:
[ADD] Article 도메인 객체에 펙터리 메서드와 null validate 기능을 추가합니다

title이나 coverUrl, contents 에 null이 들어오는 경우 예외를 던져주도록 수정하였습니다.

Deocksoo committed:
[REFACTOR] update되는 Article 도메인 객체의 생성을 ArticleRepository에서 담당하도록 수정합니다

ArticleContoller에서 담당하던 도메인 객체의 생성을 ArticleRepository에서 하도록 수정하였습니다.

Deocksoo committed:
[TEST] ArticleControllerTest에서 ArticleRepository를 제거하고 api를 활용해 테스트하도록 수정합니다

1. 수정하는 과정에서 ArticleDTO의 클래스명을 ArticleDto로 변경하였습니다.
2. Acceptance Test에 rest assured를 사용하였습니다. rest assured 라이브러리를
사용하는 경우 redirect되었을 때 경로를 String 데이터로 받아올 수 있게 되어 테스트가
한층 더 수월해집니다.

Deocksoo committed:
[REFACTOR] POST, PUT 메서드를 위한 String 값들을 스프링 HttpMethod 상수로 치환합니다

Deocksoo committed:
[REFACTOR] Article 도메인 객체 생성자의 접근제어자를 private으로 변경합니다

factory 메서드인 Article.of에서 모든 생성을 담당하도록 수정하였습니다.

Deocksoo committed:
[REFACTOR] ArticleRepository의 변수명들을 변경합니다.

setup된 article의 id를 상수로 관리하도록 하지 않고, setup될 때 필드 변수로 입력받아
사용하도록 수정하였습니다.

Deocksoo commented:
말씀해주신 부분들을 모두 반영하는 것이 좋을 것 같아서 대부분 수정하였습니다!
덕분에 Controller를 어떻게 테스트할 것인지 등 모호했던 부분들이 많이 풀린 것 같아요 :)

해결이 조금 어려웠던 부분도 있었어요. 'ArticleController에서 다음 경로까지 지정해줘야 하는가'와 관련된 이슈를 비롯해 아직 아리송한 것들이 몇 가지 있는데,  해당 부분에 코멘트로 남겨두었습니다. 한 번 확인해주실 수 있을까요?

감사합니다!

kang-hyungu reviewed (approved):
피드백 반영하신 부분 확인했습니다!
질문 주신 내용에도 답변 남겼어요
다음 미션 진행하시면 적용하시면 될 내용들이라 머지 처리할게요
코드 구현하느라 수고하셨습니다!
